In [1]:
# data read Imports
import pandas as pd

# Machine Learning Imports
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# For evaluating our ML results
from sklearn import metrics

In [4]:
df=pd.read_csv('data/sample.csv')
df

female  read  write  math  pass  femalexmath
0         0    57     52    41     0            0
1         1    68     59    53     0           53
2         0    44     33    54     0            0
3         0    63     44    47     0            0
4         0    47     52    57     0            0
..      ...   ...    ...   ...   ...          ...
195       1    55     59    52     0           52
196       1    42     46    38     0           38
197       1    57     41    57     0           57
198       1    55     62    58     1           58
199       1    63     65    65     1           65

[200 rows x 6 columns]

In [5]:
feature=df[['female','math']]
target=df['pass']


In [6]:
# help(train_test_split)
# 80 and 20 split 
X_train, X_test, y_train, y_test = train_test_split(feature, target,
                                                    test_size=0.2,
                                                    random_state=42)

In [7]:
X_train.shape, X_test.shape

((160, 2), (40, 2))

In [8]:
# scikit learn
lr = LogisticRegression()
lr.fit(X_train, y_train)

print("coef : ", lr.coef_ , " \nintercept: ", lr.intercept_)

coef :  [[0.69616515 0.16474612]]  
intercept:  [-10.6640414]


In [10]:
target.value_counts()

0    151
1     49
Name: pass, dtype: int64

In [11]:
import math
print("female : ", math.exp(lr.coef_[0][0])," math :",math.exp(lr.coef_[0][1])," intercept:" , int(math.exp(lr.intercept_[0])))

female :  2.006045061281207  math : 1.1790937366213623  intercept: 0


In [12]:
df.groupby('female')['pass'].value_counts()

female  pass
0       0       74
        1       17
1       0       77
        1       32
Name: pass, dtype: int64

### Task

In [14]:
feature=df[['female','read','write','math']]
target=df['pass']

In [15]:
# 80 and 20 split 
X_train, X_test, y_train, y_test = train_test_split(feature, target,
                                                    test_size=0.2,
                                                    random_state=42)

In [16]:
# scikit learn
lr = LogisticRegression()
lr.fit(X_train, y_train)

print("coef : ", lr.coef_ , " \nintercept: ", lr.intercept_)

coef :  [[-1.69349625e-02 -3.34269429e-02  2.35458689e+00  2.34839570e-03]]  
intercept:  [-140.75837151]


In [17]:
import math
print("female : ", math.exp(lr.coef_[0][0])," read :",math.exp(lr.coef_[0][1])," write :",math.exp(lr.coef_[0][2]),"math :",math.exp(lr.coef_[0][3])," intercept:" , int(math.exp(lr.intercept_[0])))

female :  0.9832076279518256  read : 0.9671255640815817  write : 10.533776308028976 math : 1.0023511553360211  intercept: 0


## Interpretation

If the gender is female then the odds of passing exam is 0.98. When other value fixed.
If the read number increase by one unit then the odds of passing exam is 0.96. When other value fixed.
If the write number increase by one unit then the odds of passing exam is 10. When other value fixed.
If the math number increase by one unit then the odds of passing exam is 1.00. When other value fixed.
When all numerical features are zero and the categorical features are at the reference category, the estimated odds are 0 . The interpretation of the intercept weight is usually not relevant.